### PR TITLE
UCS/async: unexpose async datatype to user level

### PR DESCRIPTION
**Pull request by shssf**
Avoid exposing ucs_async_context_t to user level. 
This PR is related to #1285 and #1233 

If changes in `gtests` and `ucx_perftest` are unlikely I can change it by using internal UCX headers to avoid many changes there.

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/876/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3344/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2839/ for details (Mellanox internal link).

File: src/ucp/core/ucp_worker.h [hunk 1]
```diff
@@ -81,7 +81,7 @@ typedef struct ucp_worker_wakeup {
  * UCP worker (thread context).
  */
 typedef struct ucp_worker {
-    ucs_async_context_t           async;         /* Async context for this worker */
+    ucs_async_context_t          *async;         /* Async context for this worker */
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.h]: need to keep async as value and not pointer for best performance

**Comment by shssf** [comment]
@yosefe In case of no `ucp_worker_t` changes allowed I implement two functions.
Internally, following functions should be used:
```
ucs_async_context_init_internal()
ucs_async_context_cleanup_internal()
```
Nothing changed in public interface but the `ucs_async_context_t` became opaque:
```
ucs_async_context_init()
ucs_async_context_cleanup()
```

**Comment by shssf** [comment]
@brminich I would appreciate for any comments

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3354/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/885/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2849/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2850/ for details (Mellanox internal link).

File: src/ucs/async/async_fwd.h [hunk 2]
```diff
@@ -1,12 +1,110 @@
...
+
+
+/**
+ * Allocate and initialize an asynchronous execution context.
+ * This can be used to ensure safe event delivery.
+ *
+ * @param async           Event context pointer to initialize.
+ * @param mode            Either to use signals or epoll threads to wait.
+ *
+ * @return Success status.
```
Comment (brminich) [inline src/ucs/async/async_fwd.h]: minor: @return tag is present here while it is missing in the other calls returning ucs_status_t

File: src/ucs/async/async_fwd.h [hunk 3]
```diff
@@ -1,12 +1,110 @@
...
+ */
+ucs_status_t ucs_async_context_init(ucs_async_context_t **async,
+                                    ucs_async_mode_t mode);
+
+
+/**
+ * Clean up the async context, and release system resources if possible.
+ * The context memory released.
+ *
+ * @param event           Asynchronous context to clean up.
```
Comment (brminich) [inline src/ucs/async/async_fwd.h]: the actual param name is async

**Comment by shssf** [comment]
Comments addressed. Fixes #1294 and `uct_hello_world` added to the documentation index.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3361/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/894/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2857/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/897/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3364/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2860/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2863/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2865/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3374/ for details.

File: src/ucs/async/async_fwd.h [hunk 4]
```diff
@@ -1,12 +1,119 @@
...
+ *
+ * Allocate and initialize an asynchronous execution context.
+ * This can be used to ensure safe event delivery.
+ *
+ * @param async           Event context pointer to initialize.
+ * @param mode            Either to use signals or epoll threads to wait.
+ *
+ * @return Error code as defined by @ref ucs_status_t.
+ */
+ucs_status_t ucs_async_context_init(ucs_async_context_t **async,
```
Comment (yosefe) [inline src/ucs/async/async_fwd.h]: 1. usually the result is the last argument and called _p. e.g.
`+ucs_status_t ucs_async_context_init(ucs_async_mode_t mode, ucs_async_context_t **async_p)`
1. what do you think about creating new functions called ucp_async_create/destroy, and keeping the previous functions ucs_async_init/cleanup as-is, for internal use only?

**Comment by yosefe** [comment]
@shamisp pls take a look on the doxygen updates for ucp.h

**Comment by shamisp** [comment]
I noticed that 3 addition directories where added to the doxygen:
```
 $(SRCDIR)/src/ucs/async/ \
 $(SRCDIR)/src/ucs/time/  \
 $(SRCDIR)/src/ucs/sys/   \
```

Can you please clarify the reason ?

**Comment by shssf** [comment]
@shamisp THese directories needs to be added because I added 3 more files into documentation index. These files contains functions/types declarations used in public interface. For example, the return type `ucs_status_t` in file `src/ucs/type/status.h` added before. Newly added files brings type declarations, into documentation, for `ucs_async_context_t`, `ucs_time_t`, UCS_BIT, etc.

File: src/ucs/async/async_fwd.h (hunk 4 above)
Comment (shssf) [inline src/ucs/async/async_fwd.h]: 1. sure, i will change it if we find a conclusion for >2<
2. I think it broke the UCS public interface. I kept function `ucs_async_context_init` as is for public and create `ucs_async_context_init_internal` for leave any public interface untouchable. For example, in `uct_hello_world` we used `ucs_async_context_init`. I assume, unknown number of external users have this function in their programs. If we replace `ucs_async_context_init` by `ucp_async_create` they have to change their programs.
In other hand, using `ucs_async_context_init_internal` isn't comfortable for UCX internally.
Do you agree to change UCS public interface?

Also, I would like to ask @shamisp opinion about this. The main goal for all these changes is to move `ucs_async_context_t` (and many related types) from user level to internal and leave this type as opaque for the user. The problem - we have to allocate memory for this type since the user have no information about the type length.

From my point of view, changing UCS public interface, is reasonable for this particular case.

Comment (yosefe) [inline src/ucs/async/async_fwd.h]: i think we can change api, since no one could include async.h anyway before this PR

**Comment by shamisp** [comment]
@shssf - can  you please provide a full least of files and specific C structures/functions that you want to document ? We would have to go one by one and make sure that everything is well documented.

File: doc/doxygen/ucxdox [hunk 5]
```diff
@@ -787,11 +790,14 @@ INPUT_ENCODING         = UTF-8
...
-                         uct_def.h  \
-                         status.h   \
+FILE_PATTERNS          = ucp.h          \
+                         ucp_def.h      \
+                         uct.h          \
+                         uct_def.h      \
+                         status.h       \
+                         async_fwd.h    \
+                         compiler_def.h \
+                         time_def.h     \
```
Comment (shssf) [inline doc/doxygen/ucxdox]: @shamisp I added these three files (time_def.h, compiler_def.h, async_fwd.h) to the documentation index because they are in public interface. I think all functions/types should be well documented from these files.

File: src/ucs/async/async.h [hunk 6]
```diff
@@ -12,8 +12,8 @@
 #include "async_fwd.h"
 
 #include <ucs/datastruct/mpmc.h>
+#include <ucs/time/time.h>
```
Comment (yosefe) [inline src/ucs/async/async.h]: why is this needed? is anything from time.h used in async.h?

Comment (yosefe) [inline src/ucs/async/async.h]: never mind.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3380/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/913/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3381/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/914/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3382/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/915/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2874/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2875/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2876/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2877/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
it was merged a bit early. We would have to revert it and get all approvals. It add news documentation to the public document that has not been reviewed.
